### ISSUE TITLE
Add function to the msgpack-rpc-api to call arbitrary functions more conveniently

### ISSUE THREAD

**Issue by saep** [neovim/neovim#2972]
This is a mere suggestion and not urgent at all.

If I want to call arbitrary function on the neovim side, I currently use `vim_command` or `vim_command_output`. 

``` yaml
- can_fail: true
  deferred: true
  name: vim_command
  parameters:
  - [String, str]
  return_type: void

- can_fail: true
  deferred: false
  name: vim_command_output
  parameters:
  - [String, str]
  return_type: String
```

In a statically typed language, it involves quite some overhead to create a function call for that. I have to manually create a valid vimL string from the types I have. Since I already have mechanisms to convert my values into msgpack-rpc objects, wouldn't it be convenient to just use these directly? I have a function like this in mind.

``` yaml
- can_fail: true
  deferred: false
  name: vim_function
  parameters:
  - [String, functionName]
  - [ArrayOf(Object), args]
  return_type: Object
```

I can achieve a similar behaviour with `vim_command_output` by creating higher order functions (or whatever) in the language I am writing a plugin in. But that is quite some overhead and another serialization I have to implement even though the msgpack-rpc serialization is already there. There may still be use-cases where you have to create vimL by hand, but I think this is one that deserves an API function.

As a motivating example, take the function `setqflist`. It takes a dictionary with a few mandatory values. If I already have a dictionary in my plugin language, then I can simply use a function like `toMSGPack` to convert it to an argument for `vim_function`. Otherwise, I have to create a string representation from the map that is valid vimL.

The only downside I see with this is that there is one more function in the API.

One noteworthy upside is that you only have to do this work once on the neovim side and not necessarily in all languages that plugins can be written in. There may even be language dependent advantages for doing this (e.g. static guarantees for argument values).

I would start creating a pull request if this is deemed to be useful.

**Comment by splinterofchaos** [comment]
Sounds good to me. This will also be more efficient on the nvim side since we will not have to parse the string and build the arguments, but rather translate them from msgpack to vimL, and we already have a function to do that.

Why do you use `vim_command` instead of `vim_eval`? It is functions you want to execute, not commands, correct?

Just a note, I believe that anything that executes vimL code should be deferred.

**Comment by justinmk** [comment]
@splinterofchaos In case you are going to work on this, just note that we are in a feature-freeze for the first release. So we want to spend most of our time fixing bugs and finishing the bare-minimum of features to get to a stable release (target date is 1 September).

So I'll add this feature to 0.2 milestone.

**Comment by fwalch** [comment]
@justinmk Did I miss an announcement?

**Comment by splinterofchaos** [comment]
@justinmk, exciting news! 

Actually, @saep said s/he'd make the PR.

**Comment by bfredl** [comment]
I also wanted this :+1:
A temporary workaround is otherwise to store the big parameter in a `g:` variable
```
vim_set_var('_myplugin_temp', biglistofcompletions)
vim_eval('complete(' + start + ', g:_myplugin_temp)')
```

**Comment by justinmk** [comment]
> Did I miss an announcement?

No I'm waiting on @tarruda before announcing specific details. But you can see the milestone date on GitHub :) https://github.com/neovim/neovim/milestones

**Comment by fwalch** [comment]
@justinmk Oh, didn't know there were dates on the milestones :-)

**Comment by tarruda** [comment]
For anyone who will work on this, here's the relevant function that needs to be exposed via msgpack: https://github.com/neovim/neovim/blob/master/src/nvim/eval.c#L7561.

Note that for the client side, funcrefs are nothing but the internal function name, so there's no need to create a new Function type to be exposed via the API(it could get complicated to handle cases like functions being passed between the client and the server).

 Just convert the function arguments and pass with the internal name to `func_call`. Clients like python can create their own function types that are callable and invoke the API function behind the scenes.

**Comment by justinmk** [comment]
closed by https://github.com/neovim/neovim/pull/2979
